channels/api.py

```diff
@@ -56,26 +69,97 @@ def get_or_create_user(username):
     refresh_token_url = urljoin(settings.OPEN_DISCUSSIONS_REDDIT_URL, '/api/v1/generate_refresh_token')
 
     session = _get_session()
-    resp = session.get(refresh_token_url, params={'username': username}).json()
-    return resp['refresh_token']
+    return session.get(refresh_token_url, params={'username': username}).json()
+
 
+def get_auth_tokens(user):
+    """
+    Gets the stored refresh token or generates a new one
 
-def _get_user_credentials(user):
+    Args:
+        user (User): the authenticated user
+
+    Returns:
+        (channels.models.RedditRefreshToken, channels.models.RedditAccessToken): the stored tokens
     """
-    Get credentials for authenticated user
+    threshold_date = datetime.now(pytz.utc) + timedelta(minutes=2)
+    refresh_token, _ = RedditRefreshToken.objects.get_or_create(user=user)
+    access_token = None
+
+    # if we created this token just now, atomically generate one
+    if not refresh_token.token_value:
+        with transaction.atomic():
+            refresh_token = RedditRefreshToken.objects.filter(user=user).select_for_update()[0]
+            if not refresh_token.token_value:
+                response = _get_refresh_token(user.username)
+                refresh_token.token_value = response['refresh_token']
+                refresh_token.save()
+
+                # the response also returns a valid access_token, so we might as well store that for use
+                # offset it negatively a bit to account for response time
+                expires_at = datetime.now(pytz.utc) + timedelta(seconds=response['expires_in'] - EXPIRES_IN_OFFSET)
+                access_token = RedditAccessToken.objects.create(
+                    user=user,
+                    token_value=response['access_token'],
+                    token_expires_at=expires_at.replace(tzinfo=pytz.utc)
+                )
+
+    # return the refresh token and access_token
+    return refresh_token, access_token or RedditAccessToken.valid_tokens_for_user(user, threshold_date).first()
```

Comment:
Actually is there any case where we shouldn't just use `RedditAccessToken.valid_tokens_for_user`?

Reply:
Actually, I'm wondering if we should just return only `refresh_token` here and do the `RedditAccessToken.valid_tokens_for_user` call in `_configure_access_token`, instead of having it be passed in as an argument. What do you think?

Reply:
I was trying to avoid an extra trip to the database, but if you don't think it's necessary I can refactor.

Reply:
I'm fine to leave it as is